pkg/controller/registry/resolver/solver/dict.go

```diff
@@ -191,36 +187,69 @@ func compileDict(installables []Installable) *dict {
 		}
 	}
 
-	d.cards = d.c.CardSort(weights)
+	d.cards = d.c.CardSort(ms)
 
 	return &d
 }
 
-// Solve searches for a solution to the problem encoded in a dict that
-// minimizes the total weight of all selected Installables.
-func (d *dict) Solve(ctx context.Context, g *gini.Gini) int {
+func (d *dict) ApplyConstraints(g *gini.Gini) {
+	ms := make([]z.Lit, 0, len(d.lits))
+	for _, m := range d.lits {
+		ms = append(ms, m)
+	}
 	d.c.ToCnf(g)
+	for m := range d.constraints {
+		g.Assume(m)
+	}
+}
 
-	var result int
-	linearSearch(0, d.cards.N(), func(w int) bool {
-		if w >= 0 {
-			// Omitting the cardinality constraint on the
-			// last iteration prevents the cardinality
-			// constraint literal from showing up in the
-			// set of failed assumptions should there be
-			// no solution.
-			g.Assume(d.cards.Leq(w))
+func (d *dict) MinimizeCardinality(g *gini.Gini) error {
+	for w := 0; w <= d.cards.N(); w++ {
+		g.Assume(d.cards.Leq(w))
```

Comment:
If we're spending a lot of time here, can we make an argument for a minimum value that's greater than 0?

The cardinality of the solution should be at least the length of all mandatory identifiers, right?

Reply:
Most of the cycles were spent building the sorting network in `compileDict`. I'm pushing up a change that postpones building the sorting network until after a solution is found, and had really good results.

```
// before
BenchmarkSolve-8   	     216	  54731543 ns/op
// after
BenchmarkSolve-8   	    9351	   1234811 ns/op
```

I updated BenchmarkSolve as well so that it covers both `compileDict` and `Solve` together. I think that's more realistic.